rubix-bookkeeper/src/main/java/com/qubole/rubix/bookkeeper/LocalDataTransferServer.java

```diff
@@ -217,52 +217,55 @@ public void stop()
+          dataInfo.flip();
+
+          DataTransferHeader header = DataTransferClientHelper.readHeaders(dataInfo);
+          long offset = header.getOffset();
+          int readLength = header.getReadLength();
+          String remotePath = header.getFilePath();
+          log.info(String.format("Trying to read from %s at offset %d and length %d for client %s", remotePath, offset, readLength, localDataTransferClient.getRemoteAddress()));
+          try (RetryingPooledBookkeeperClient bookKeeperClient = bookKeeperFactory.createBookKeeperClient(conf)) {
+            if (!CacheConfig.isParallelWarmupEnabled(conf)) {
+              if (!bookKeeperClient.readData(remotePath, offset, readLength, header.getFileSize(),
+                      header.getLastModified(), header.getClusterType())) {
+                throw new Exception("Could not cache data required by non-local node");
+              }
+            }
+            else {
+              // Just make sure the requested blocks are present in the cache. If any the blocks is
+              // not presnt in the cache, throw an exception so that the caller NonLocalReadRequestChain
+              // can read the data from the object store
+              long blockSize = CacheConfig.getBlockSize(conf);
+              long startBlock = offset / blockSize;
+              long endBlock = ((offset + (readLength - 1)) / blockSize) + 1;
+
+              CacheStatusRequest request = new CacheStatusRequest(remotePath, header.getFileSize(), header.getLastModified(),
+                      startBlock, endBlock).setClusterType(header.getClusterType());
+              List<BlockLocation> blockLocations = bookKeeperClient.getCacheStatus(request);
+
+              long blockNum = startBlock;
+              for (BlockLocation location : blockLocations) {
+                if (location.getLocation() != Location.CACHED) {
+                  log.error(String.format("The requested data for block %d of file %s in not in cache. " +
+                          " The data will be read from object store", blockNum, remotePath));
+                  throw new Exception("The requested data in not in cache. The data will be read from object store");
+                }
               }
               blockNum++;
             }
+            int nread = readDataFromCachedFile(bookKeeperClient, remotePath, offset, readLength);
+            log.info(String.format("Done reading %d from %s at offset %d and length %d for client %s", nread, remotePath, offset, readLength, localDataTransferClient.getRemoteAddress()));
           }
-
-          int nread = readDataFromCachedFile(bookKeeperClient, remotePath, offset, readLength);
-          log.debug(String.format("Done reading %d from %s at offset %d and length %d for client %s", nread, remotePath, offset, readLength, localDataTransferClient.getRemoteAddress()));
         }
       }
       catch (Exception e) {
```

Comment:
I have added debug log at Link: https://github.com/qubole/rubix/pull/390/files#diff-b216de043d663aa296b0aa2e067febcfR235. Here, we get an exception in case of a scavenger closing. I will remove the debug logs too.